### DISCUSSION TITLE
Workaround for playback in Safari?

### DISCUSSION THREAD

**Discussion by moenodedev** [bluenviron/mediamtx#3359] [Questions]
### Question

Does anyone have a workaround to make recordings play from a `<video>` element in Safari? Is there a JavaScript library that can play the recording? We're running an `ffmpeg` command with the `runOnRecordSegmentComplete` hook to fix the fmp4 segments but that's inefficient.

**Comment by moenodedev** [comment]
#3350 is the bug

**Comment by bouk** [comment]
@moenodedev could you share the ffmpeg command you're using with runOnRecordSegmentComplete to fix the mp4 segments?

**Reply by moenodedev** [reply]
The `runOnRecordSegmentComplete` hook moves the segment to $filename.original, fixes the segment with ffmpeg, gets the duration with ffprobe, writes the filename and duration to segments.txt, then deletes the broken original.

runOnRecordSegmentComplete shell commands:
```sh
mv $MTX_SEGMENT_PATH $MTX_SEGMENT_PATH.original \
  && ffmpeg -i $MTX_SEGMENT_PATH.original -c copy $MTX_SEGMENT_PATH \
  && echo $MTX_SEGMENT_PATH $(ffprobe -i $MTX_SEGMENT_PATH -show_entries format=duration -v quiet -of csv=p=0) >>/recordings/$MTX_PATH/segments.txt \
  && rm $MTX_SEGMENT_PATH.original
```

runOnRecordSegmentComplete configuration (returned from the API):
```json
{
  "runOnRecordSegmentComplete": "sh -c \"mv $MTX_SEGMENT_PATH $MTX_SEGMENT_PATH.original && ffmpeg -i $MTX_SEGMENT_PATH.original -c copy $MTX_SEGMENT_PATH && echo $MTX_SEGMENT_PATH $(ffprobe -i $MTX_SEGMENT_PATH -show_entries format=duration -v quiet -of csv=p=0) >>/recordings/$MTX_PATH/segments.txt && rm $MTX_SEGMENT_PATH.original\""
}
```

segments.txt
```
./recordings/skycam/2024-06-06_00-38-06-422160.ts 10.176767
./recordings/skycam/2024-06-06_00-38-16-588926.ts 10.176778
./recordings/skycam/2024-06-06_00-38-26-769670.ts 10.176822
./recordings/skycam/2024-06-06_00-38-36-940687.ts 10.176778
```

**Reply by moenodedev** [reply]
There's an open issue to make `duration` available to the `runOnRecordSegmentComplete` hook. We wouldn't need to inspect the segment with ffprobe.

https://github.com/bluenviron/mediamtx/issues/2983

**Reply by moenodedev** [reply]
The segments in `segments.txt` end in `.ts` not `.fmp4` because we're using this to process `mpegts` segments not `fmp4` segments. But the ffmpeg command works for both mpegts and fmp4.

More details are in https://github.com/bluenviron/mediamtx/discussions/3243

**Reply by aler9** [reply]
@bouk since this issue was fixed in the main branch, you can also build the server from source or wait until the next release, without performing remuxing.

[a comment by an automated account is not shown]
